sycl/test/native_cpu/scalar_args.cpp

```diff
@@ -12,39 +12,62 @@ const size_t N = 10;
 
 template <typename T> class init_a;
 
-template <typename T> bool test(queue myQueue) {
-  {
-    buffer<float, 1> a(range<1>{N});
-    const T test = rand() % 10;
-
-    myQueue.submit([&](handler &cgh) {
-      auto A = a.get_access<access::mode::write>(cgh);
-      cgh.parallel_for<init_a<T>>(range<1>{N},
-                                  [=](id<1> index) { A[index] = test; });
-    });
-
-    auto A = a.get_access<access::mode::read>();
-    std::cout << "Result:" << std::endl;
-    for (size_t i = 0; i < N; i++) {
-      if (A[i] != test) {
-        std::cout << "ERROR at pos " << i << " expected " << test << ", got "
-                  << A[i] << "\n";
-        return false;
-      }
-    }
+template <typename T, int M>
+bool compare(const sycl::vec<T, M> a, const sycl::vec<T, M> truth) {
+  bool res = true;
+  for (int i = 0; i < M; i++) {
+    res &= a[i] == truth[i];
   }
+  return res;
+}
 
-  std::cout << "Good computation!" << std::endl;
+template <typename T> bool compare(const T a, const T truth) {
+  return a == truth;
+}
+
+template <typename T> bool check(buffer<T, 1> &result, const T truth) {
+  auto A = result.get_host_access();
+  for (size_t i = 0; i < N; i++) {
+    if (!compare(A[i], truth)) {
+      return false;
+    }
+  }
   return true;
 }
 
+template <typename T> bool test(queue myQueue) {
+  buffer<T, 1> a(range<1>{N});
+  const T test{42};
+
+  myQueue.submit([&](handler &cgh) {
+    auto A = a.template get_access<access::mode::write>(cgh);
+    cgh.parallel_for<init_a<T>>(range<1>{N},
+                                [=](id<1> index) { A[index] = test; });
+  });
+
+  return check(a, test);
+}
+
 int main() {
   queue q;
-  int res1 = test<int>(q);
-  int res2 = test<unsigned>(q);
-  int res3 = test<float>(q);
-  int res4 = test<double>(q);
-  if (!(res1 && res2 && res3 && res4)) {
+
+  std::vector<bool> res;
+  res.push_back(test<int>(q));
+  res.push_back(test<unsigned>(q));
```

Comment:
Performance. Using a vector here likely causes allocations and there's the overhead of iterating over the vector. 
It's probably faster to instead check the test values and return straight away if there's an error.